### PR TITLE
[FEATURE] add score_matrix and trace_matrix for edit_distance

### PR DESCRIPTION
**Pull request by marehr**

### CODE REVIEW

File: test/unit/alignment/matrix/edit_distance_trace_matrix_full_test.cpp [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+
+    EXPECT_EQ(result, expect);
+}
+
+TEST(global, multiple_words)
+{
+    matrix_type<false, false> matrix{18u};
+    matrix.reserve(10u);
+
+    matrix.add_column({0b0000'0000u, 0b0000'0000u, 0b0u}, {0b0000'0000u, 0b0000'0000u, 0b0u}, {0b1111'1111u, 0b1111'1111u, 0b1u});
```
Comment (eseiler) [inline test/unit/alignment/matrix/edit_distance_trace_matrix_full_test.cpp]: I'd say we enforce line lengths here

File: include/seqan3/alignment/matrix/edit_distance_score_matrix_full.hpp [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+        score_type const p = bitset(vp).count();
+        score_type const n = bitset(vn).count();
+
+        return p - n;
+    }
+
+public:
+    //!\copydoc seqan3::detail::Matrix::at
+    entry_type at(size_t const row, size_t const col) const noexcept
```
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/alignment/matrix/edit_distance_score_matrix_full.hpp]: Good question, I don't know! What would be the exception type?

File: test/unit/alignment/matrix/edit_distance_score_matrix_full_test.cpp [hunk 3]
```diff
@@ -0,0 +1,715 @@
...
+
+using score_type = int;
+using word_type = uint8_t;
+
+template <bool is_semi_global, bool use_max_errors>
+class matrix_type
+    : public seqan3::detail::edit_distance_score_matrix_full<word_type, score_type, is_semi_global, use_max_errors>
+{
+public:
+    using base_t = seqan3::detail::edit_distance_score_matrix_full<word_type, score_type, is_semi_global, use_max_errors>;
```
[the remark this answers is not in the document]
Comment (marehr) [reply test/unit/alignment/matrix/edit_distance_score_matrix_full_test.cpp]: no 122 :laughing:

File: include/seqan3/alignment/matrix/edit_distance_score_matrix_full.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (eseiler) [reply include/seqan3/alignment/matrix/edit_distance_score_matrix_full.hpp]: [`std::out_of_range`](https://en.cppreference.com/w/cpp/error/out_of_range)?
But it's a general design decision and would need to change in multiple places, so it's out of scope for this PR.

File: include/seqan3/alignment/matrix/edit_distance_trace_matrix_full.hpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+        return columns.size();
+    }
+
+protected:
+    //!\brief If use_max_errors is true store these additional state information in state_type.
+    struct max_errors_state
+    {
+        //!\brief The number of max_rows within the current column. Computed by
+        //!       seqan3::detail::edit_distance_score_matrix_full::max_rows.
+        unsigned max_rows{};
```
Comment (eseiler) [inline include/seqan3/alignment/matrix/edit_distance_trace_matrix_full.hpp]: not fixed length on purpose?

Comment (marehr) [reply include/seqan3/alignment/matrix/edit_distance_trace_matrix_full.hpp]: Sorry I don't understand the comment.

Comment (marehr) [reply include/seqan3/alignment/matrix/edit_distance_trace_matrix_full.hpp]: I changed it to `size_t`

File: test/unit/alignment/matrix/edit_distance_score_matrix_full_test.cpp [hunk 5]
```diff
@@ -0,0 +1,716 @@
...
+
+    matrix_type(size_t const rows_size) : base_t{rows_size}
+    {}
+
+    using base_t::max_rows;
+    using base_t::add_column;
+    using base_t::reserve;
+};
+
+static constexpr score_type INF = 2147483647;
```
Comment (rrahn) [inline test/unit/alignment/matrix/edit_distance_score_matrix_full_test.cpp]: Should this constant not be exposed by the matrix types?

File: include/seqan3/alignment/matrix/edit_distance_score_matrix_full.hpp [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    static score_type score_delta_of_word(word_type const & vp, word_type const & vn) noexcept
+    {
+        score_type const p = std::bitset<word_size>{vp}.count();
+        score_type const n = std::bitset<word_size>{vn}.count();
+        return p - n;
+    }
+
+public:
+    //!\copydoc seqan3::detail::Matrix::at
+    entry_type at(size_t const row, size_t const col) const noexcept
```
[the remark this answers is not in the document]
Comment (marehr) [reply include/seqan3/alignment/matrix/edit_distance_score_matrix_full.hpp]: in a later PR

File: include/seqan3/alignment/matrix/edit_distance_score_matrix_full.hpp [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    /*!\brief Computes the number of max rows in the current column.
+     * \tparam score_type \copybrief score_type
+     * \param score_mask  \copybrief pairwise_alignment_edit_distance_unbanded::score_mask
+     * \param last_block  \copybrief pairwise_alignment_edit_distance_unbanded::last_block
+     * \param score       \copybrief pairwise_alignment_edit_distance_unbanded::_score
+     * \param max_errors  \copybrief pairwise_alignment_edit_distance_unbanded::max_errors
+     * \return Number of max rows in the current column.
+     */
+    template <typename score_type>
+    static size_t max_rows(word_type const score_mask, unsigned const last_block,
```
Comment (rrahn) [inline include/seqan3/alignment/matrix/edit_distance_score_matrix_full.hpp]: is this not noexcept?

File: include/seqan3/alignment/matrix/edit_distance_score_matrix_full.hpp [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    //!\copydoc seqan3::detail::Matrix::entry_type
+    using entry_type = std::conditional_t<use_max_errors, std::optional<score_type>, score_type>;
+
+    //!\brief A special score which represents infinity.
+    static constexpr std::optional<score_type> inf = std::nullopt;
+
+    /*!\brief Increase the capacity of the columns to a value that is greater or equal to `new_capacity`.
+     * \param new_capacity The new capacity.
+     */
+    void reserve(size_t const new_capacity)
```
Comment (rrahn) [inline include/seqan3/alignment/matrix/edit_distance_score_matrix_full.hpp]: can you add a comment about exception safety here

**Comment by marehr** [comment]
@rrahn ping

**Pull request merged**